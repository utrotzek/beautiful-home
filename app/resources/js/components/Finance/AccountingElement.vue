<template>
    <div class="accounting-element row mb-3">
        <div class="col accounting-element-inner">
            <div class="delete-element">
                <button>
                    <i class="fa fa-trash-alt"></i>
                </button>
            </div>

            <div class="row">
                <div class="date col-12">
                    {{ accountingDate }}
                </div>
            </div>
            <div class="row">
                <div class="accounting-title col-8">
                    <span
                        v-if="showCheck"
                        class="fa fa-check-circle check"
                    ></span>
                    {{ accountingTitle }}
                </div>
                <div
                    class="remaining-amount col-4"
                    :class="classObject"
                >
                    <span v-if="!showCheck">{{ remainingAmount }} â‚¬</span>
                </div>
            </div>

            <div class="row relatedElements">
                <div
                    v-for="(planningElement) in planningElements"
                    :key="planningElement.id"
                    class="col-12 col-md-6"
                >
                    <PlanningElement
                        :id="planningElement.id"
                        is-connected
                        :has-date="false"
                        :description="planningElement.description"
                        :title="planningElement.title"
                        :total-amount="planningElement.totalAmount"
                        :date="planningElement.date"
                        @delete="deletePlanning"
                    />
                </div>
            </div>
        </div>
    </div>
</template>

<script>

import PlanningElement from "../../../js/components/Finance/PlanningElement";

export default {
    name: "AccountingElementVue",

    components: {
        PlanningElement
    },
    props: {
        totalAmount: {
            type: Number,
            required: true
        },

        accountingTitle: {
            type: String,
            required: true
        },

        accountingDate: {
            type: String,
            required: true
        },
    },
    data() {
        return {
            planningElements: [
                {
                    id: 1,
                    totalAmount: -1,
                    title: "Einkaufen",
                    description: "this is my descipriotn",
                    date: "20.10.2019"
                },
                {
                    id: 2,
                    totalAmount: -99,
                    title: "Amazon",
                    description: "Weil das so ist",
                    date: "20.10.2019"
                }
            ],
        };
    },
    computed: {
        remainingAmount() {
            let i = 0,
                remainingAmmount = this.totalAmount;

            for (i  = 0; i < this.planningElements.length; i++){
                remainingAmmount = remainingAmmount - this.planningElements[i].totalAmount;
            }
            return remainingAmmount;
        },
        showCheck() {
            return this.remainingAmount === 0;
        },
        classObject() {
            return {
                "negative": this.remainingAmount < 0,
                "positive": this.remainingAmount >= 0
            };
        }
    },
    methods: {
        removeFromArray(arrayList, id){
            return arrayList.filter(function(ele){
                return ele.id !== id;
            });
        },
        deletePlanning(id) {
            this.planningElements = this.removeFromArray(this.planningElements, id);
        }
    }
};
</script>

<style scoped>
    .accounting-element-inner {
        border: solid #adb5bd 1px;
        padding: 10px 15px 10px 10px;
    }

    .delete-element{
        position: absolute;
        right: 8px;
        top: 2px;
        z-index: 100;
    }

    .delete-element button {
        background-color: transparent;
        border:none;
    }

    .remaining-amount {
        text-align: right;
    }

    .date {
        font-size: 80%;
    }

    .relatedElements{
        margin-top: 1rem;
    }

    .positive,
    .check {
        color: green;
    }

    .negative {
        color: red;
    }
</style>
