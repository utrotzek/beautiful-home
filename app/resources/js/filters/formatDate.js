import moment from "moment";

//TODO: set format in moment
export default function formatDate (value) {
    if (value) {
        if (typeof value === "object"){
            return moment(value.date).format("DD.MM.YYYY");
        }else{
            return moment(String(value)).format("DD.MM.YYYY");
        }
    }
}
