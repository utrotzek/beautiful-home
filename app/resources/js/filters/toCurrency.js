export default function toCurrency (value) {
    if (typeof value !== "number") {
        return value;
    }
    var formatter = new Intl.NumberFormat("de-DE", {
        style: "currency",
        currency: "EUR",
        minimumFractionDigits: 2
    });
    return formatter.format(value);
}
