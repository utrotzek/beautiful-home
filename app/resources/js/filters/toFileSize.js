export default function toFileSize(value){
    if (isNaN(value)){
        return "undefined size";
    }else{
        const units = [ "byte", "kB", "MB", "GB", "TB"];
        let newValue = value;
        let counter = 0;

        if (newValue >= 1000){
            while (newValue >= 1000){
                newValue = newValue / 1024;
                counter++;
            }
        }

        var formatter = new Intl.NumberFormat("de-DE", {
            minimumFractionDigits: 1,
            maximumFractionDigits: 1
        });
        return formatter.format(newValue) + " " + units[counter];
    }
}
