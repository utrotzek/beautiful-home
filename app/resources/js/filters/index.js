import toCurrency from "./toCurrency";
import formatDate from "./formatDate";
import toFileSize from "./toFileSize";

export default {
    install(Vue) {
        Vue.filter("toCurrency", toCurrency);
        Vue.filter("formatDate", formatDate);
        Vue.filter("toFileSize", toFileSize);
    }
};
