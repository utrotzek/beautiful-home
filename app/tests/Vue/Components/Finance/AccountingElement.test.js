import "../Mocks/MatchMedia.mock";
import _ from "lodash";
import {mount} from "@vue/test-utils";
import AccountingElement from "../../../../resources/js/components/Finance/AccountingElement";
import Vue from "vue";
import Vuelidate from "vuelidate";
import filters from "../../../../resources/js/filters";
import { vueTopprogress } from "vue-top-progress";
import VCalendar from "v-calendar";
import AccountingFixtures from "./Fixtures/AccountingFixtures";
import moment from "moment";

Vue.component("Progressbar", vueTopprogress);
Vue.use(filters);
Vue.use(Vuelidate);
Vue.use(VCalendar);

describe("Accounting element", () => {
    let wrapper = null;
    let accountingData = null;

    beforeEach(() => {
        //TODO: fix this
        //clone is a workaround because the component is unfortunately manipulating the properties
        //directly (which is wrong)
        accountingData = _.cloneDeep(AccountingFixtures[0]);

        wrapper = mount(AccountingElement,{
            propsData: {
                year: 3000,
                month: 1,
                showConnectTarget: false,
                accountingData: accountingData
            }
        });
    });

    it ("renders properly", () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    it ("renders properly in edit mode", async () => {
        await enableEditMode();
        expect(wrapper.element).toMatchSnapshot();
    });

    it ("will have empty input field instead of 0 if the amount becomes zero", async () => {
        wrapper.vm.localAccountingData.totalAmount = "0";
        await enableEditMode();
        expect(wrapper.find(".amount input").element.value).toEqual("");
    });

    it("remaining amount will be synced with total amount as long it is a new item", async () => {
        wrapper.vm.localAccountingData.isNew = true;

        await enableEditMode();
        wrapper.vm.localAccountingData.totalAmount = -300.40;
        await Vue.nextTick();
        expect(wrapper.vm.localAccountingData.remainingAmount).toEqual(-300.40);

        wrapper.vm.localAccountingData.totalAmount = -400.20;
        await Vue.nextTick();
        expect(wrapper.vm.localAccountingData.remainingAmount).toEqual(-400.20);

    });

    it ("can switch to edit mode", async () => {
        expect(wrapper.vm.localAccountingData.editMode).toBeFalsy();

        wrapper.find(".action-buttons .edit").trigger("click");
        await Vue.nextTick();

        expect(wrapper.vm.localAccountingData.editMode).toBeTruthy();
    });

    it ("will fire delete event when delete button was clicked", async () => {
        wrapper.find(".action-buttons .delete").trigger("click");
        await Vue.nextTick();
        expect(wrapper.emitted().deleteAccounting[0]).toEqual([accountingData]);
    });

    it ("calculates start date properly", () => {
        const expectedDate = "1.1.3000";
        const date = wrapper.vm.startDate;
        const actualDate = date.getDate() + "." + (date.getMonth()+1) + "." + date.getFullYear();
        expect(actualDate).toEqual(expectedDate);
    });

    it ("calculates end date properly", () => {
        const expectedDate = "31.1.3000";
        const date = wrapper.vm.endDate;
        const actualDate = date.getDate() + "." + (date.getMonth()+1) + "." + date.getFullYear();
        expect(actualDate).toEqual(expectedDate);
    });

    it ("will validate totalAmount properly", async() => {
        await enableEditMode();
        const amountSelector = ".remaining-amount input";

        //set form data
        wrapper.find(amountSelector).setValue("");
        await Vue.nextTick();

        //assertions
        expect(wrapper.find(amountSelector).classes()).toContain("is-invalid");
        expect(wrapper.find(".save").attributes("disabled")).toEqual("disabled");
    });

    it ("will validate title properly", async() => {
        await enableEditMode();
        const titleSelector = ".accounting-title textarea";

        //set form data
        wrapper.find(titleSelector).setValue("");
        await Vue.nextTick();

        //assertions
        expect(wrapper.find(titleSelector).classes()).toContain("is-invalid");
        expect(wrapper.find(".save").attributes("disabled")).toEqual("disabled");
    });

    it ("will fire event when connection target was clicked", async () => {
        wrapper.setProps({showConnectTarget: true});
        await Vue.nextTick();

        wrapper.find(".connectTarget").trigger("click");
        await Vue.nextTick();

        expect(wrapper.emitted().doConnection[0]).toEqual([accountingData.id]);
    });

    it ("will fire event when connected planning was deleted", async  () =>{
        const connectedPlanningToDelete = accountingData.connectedPlanning[0];
        wrapper.vm.deletePlanning(connectedPlanningToDelete.id);
        expect(wrapper.emitted().deleteConnection[0]).toEqual(
            [
                accountingData.id,
                connectedPlanningToDelete
            ]
        );
    });

    it ("will fire event when connected planning was updated", async () => {
        let updatedConnectedPlanning = accountingData.connectedPlanning[0];
        let expectedUpdatedItem = _.cloneDeep(updatedConnectedPlanning);

        updatedConnectedPlanning.editMode = true;
        wrapper.vm.saveEditPlanning(updatedConnectedPlanning);

        expect(wrapper.emitted().updateConnectedPlanning[0]).toEqual(
            [
                expectedUpdatedItem,
                accountingData.id
            ]
        );
    });

    it ("can toggle edit mode for connected planning", async () => {
        const connectedPlanningToDelete = accountingData.connectedPlanning[0];

        wrapper.vm.editPlanning(connectedPlanningToDelete.id, true);
        expect(connectedPlanningToDelete.editMode).toBeTruthy();

        wrapper.vm.editPlanning(connectedPlanningToDelete.id, false);
        expect(connectedPlanningToDelete.editMode).toBeFalsy();
    });

    it ("wont fire event when form is invalid", () => {
        accountingData.title = "";
        wrapper.vm.save();
        expect(wrapper.emitted().updateData).toBeFalsy();
    });

    it ("can be edited", async () => {
        await enableEditMode();

        wrapper.vm.date = moment("03.03.3000", "DD.MM.YYYY").toDate();
        await wrapper.vm.$nextTick();

        wrapper.find(".save").trigger("click");
        await wrapper.vm.$nextTick();

        const expectedDate = moment("03.03.3000", "DD.MM.YYYY").toDate();
        const expectedData =  {
            id: 1,
            title: "Rewe",
            totalAmount: -100,
            remainingAmount: 0,
            editMode: false,
            date: expectedDate,
            connectedPlanning: [
                {
                    id: 100,
                    totalAmount: -40,
                    editMode: false,
                    costCenter: {
                        id: 1,
                        title: "Einkaufen"
                    },
                },
                {
                    id: 105,
                    totalAmount: -60,
                    editMode: false,
                    costCenter: {
                        id: 3,
                        title: "Sonstiges"
                    },
                },
            ],
        };
        expect(wrapper.emitted().updateData[0]).toEqual([expectedData]);
    });

    it ("will fire delete event if cancel was clicked and accounting is new", async () => {
        accountingData.isNew = true;
        wrapper.vm.cancel();
        expect(wrapper.emitted().deleteAccounting[0]).toEqual([accountingData]);
    });

    it ("will restore original data when cancel was clicked", async () => {
        const originalAccountingData = _.clone(accountingData);
        accountingData.title = "my changed title";
        accountingData.totalAmount = "-300";
        accountingData.date = moment("03.03.2003", "DD.MM.YYYY").toDate();
        wrapper.vm.cancel();
        expect(accountingData).toEqual(originalAccountingData);
    });

    async function enableEditMode(){
        wrapper.vm.localAccountingData.editMode = true;
        await Vue.nextTick();
    }
});
