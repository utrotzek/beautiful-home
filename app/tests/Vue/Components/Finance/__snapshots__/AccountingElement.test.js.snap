// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accounting element renders properly 1`] = `
<div
  class="accounting-element row mb-3"
>
  <div
    class="col accounting-element-inner"
  >
    <div
      class="action-buttons"
    >
      <button
        class="edit"
        title="Bearbeiten"
      >
        <i
          class="fa fa-edit"
        />
      </button>
       
      <button
        class="delete"
        title="Löschen"
      >
        <i
          class="fa fa-trash-alt"
        />
      </button>
    </div>
     
    <div
      class="row"
    >
      <div
        class="date col-9"
      >
        <span>
          
        </span>
      </div>
    </div>
     
    <div
      class="row"
    >
      <div
        class="accounting-title col-9"
      >
        <span
          class="fa fa-check-circle check"
        />
         
        <span
          class="pre-formatted"
        >
          Rewe
        </span>
      </div>
       
      <div
        class="amount remaining-amount col-3 positive"
      >
        <!---->
      </div>
    </div>
     
    <div
      class="row relatedElements"
    >
      <div
        class="col-12 col-md-6"
      >
        <div
          class="planningElement form-group show-overlay-button"
        >
          <!---->
           
          <!---->
           
          <div
            class="title cost-center"
          >
            <div
              class="float-left"
            >
              <span>
                Einkaufen
              </span>
            </div>
             
            <div
              class="float-right text-right negative"
            >
              <span>
                -€40.00
              </span>
            </div>
          </div>
           
          <div
            class="clearfix"
          />
           
          <div
            class="description"
          >
            <span
              class="pre-formatted"
            >
              
            </span>
          </div>
           
          <!---->
        </div>
      </div>
      <div
        class="col-12 col-md-6"
      >
        <div
          class="planningElement form-group show-overlay-button"
        >
          <!---->
           
          <!---->
           
          <div
            class="title cost-center"
          >
            <div
              class="float-left"
            >
              <span>
                Sonstiges
              </span>
            </div>
             
            <div
              class="float-right text-right negative"
            >
              <span>
                -€60.00
              </span>
            </div>
          </div>
           
          <div
            class="clearfix"
          />
           
          <div
            class="description"
          >
            <span
              class="pre-formatted"
            >
              
            </span>
          </div>
           
          <!---->
        </div>
      </div>
       
      <div
        class="col-12 col-md-6"
      >
        <div
          class="connectPlaceholder"
        >
          <span
            class="fa align-middle"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Accounting element renders properly in edit mode 1`] = `
<div
  class="accounting-element row mb-3"
>
  <div
    class="col accounting-element-inner"
  >
    <!---->
     
    <div
      class="row"
    >
      <div
        class="date col-9"
      >
        <span
          data-v-64ee1ddd=""
        >
          <input
            class="vc-appearance-none vc-text-base vc-text-gray-800 vc-bg-white vc-border vc-border-gray-400 vc-rounded vc-w-full vc-py-2 vc-px-3 vc-leading-tight focus:vc-outline-none focus:vc-shadow"
            data-v-64ee1ddd=""
            type="input"
            value="04/22/2020"
          />
          <div
            class="vc-popover-content-wrapper"
            data-v-64ee1ddd=""
            data-v-88de91da=""
          >
            <!---->
          </div>
        </span>
      </div>
    </div>
     
    <div
      class="row"
    >
      <div
        class="accounting-title col-9"
      >
        <!---->
         
        <textarea
          class="form-control"
          type="text"
        />
      </div>
       
      <div
        class="amount remaining-amount col-3 positive"
      >
        <input
          class="form-control text-right"
          type="text"
        />
      </div>
    </div>
     
    <div>
      <div
        aria-label="Basic example"
        class="btn-group d-flex"
        role="group"
      >
        <button
          class="save btn btn-light"
          type="button"
        >
          <i
            class="fas fa-save"
          />
           
          <span
            class="d-none d-sm-inline"
          >
            Speichern
          </span>
        </button>
         
        <button
          class="cancel btn btn-light"
          type="button"
        >
          <i
            class="fas fa-ban"
          />
           
          <span
            class="d-none d-sm-inline"
          >
            Abbrechen
          </span>
        </button>
         
        <!---->
         
        <!---->
         
        <!---->
         
        <!---->
      </div>
    </div>
  </div>
</div>
`;
